src/auth/Crypto.cc

```diff
@@ -35,30 +36,58 @@
 #include "common/debug.h"
 #include <errno.h>
 
-int get_random_bytes(char *buf, int len)
+// use a mersenne_twister_engine for random numbers. this engine is faster than
+// std::random_device, and generates higher quality randomness than the other
+// alternatives
+using base_random_engine = std::mt19937;
+
+// wrap with an independent_bits_engine to allow get_random_bytes() to sample
+// a single byte at a time. uniform_int_distribution can't be used for this,
+// because it is undefined for integer types smaller than short
+using streaming_random_engine = std::independent_bits_engine<
+    base_random_engine, 8, unsigned char>;
+
+// engine factory function
+static streaming_random_engine make_seeded_random_engine()
 {
-  int fd = TEMP_FAILURE_RETRY(::open("/dev/urandom", O_RDONLY));
-  if (fd < 0)
-    return -errno;
-  int ret = safe_read_exact(fd, buf, len);
-  VOID_TEMP_FAILURE_RETRY(::close(fd));
-  return ret;
+  // mt19937 has more internal state than can be seeded with a single integer.
+  // use a random_device to produce a seed sequence of 8
+  std::random_device r;
```

Comment:
i don't think we can rely on the entropy being accurate. quoting [cppreference on random_device::entropy()](http://en.cppreference.com/w/cpp/numeric/random/random_device/entropy):

> **Notes**
This function is not fully implemented in some standard libraries. For example, GNU libstdc++ and LLVM libc++ always return zero even though the device is non-deterministic (for libstdc++, see bug 67578). In comparison, Microsoft Visual C++ implementation always returns 32, and boost.random returns 10.

Reply:
Ahura H. Mazda that's infuriating. I can understand them not calculating it accurately, but at least giving a basic guidance as to whether they failed to open `/dev/urandom` and fell back to something deterministic should be a minimum.

All right, never mind that then.

Reply:
The standard says (26.5.6) that entropy() returns 0.0 if the implementation uses a random number engine, otherwise an entropy estimate. See the full entry for details of how that's calculated. (That doesn't really help with incomplete standard library implementations!) As I read it, it doesn't mean you will necessarily have poor random numbers (as in @cbodley 's example, where it actually /is/ using a non-deterministic device). Anyway, not sure that helps with anything, but maybe it's worth a peek under the covers.

Reply:
Oh! std::random_device()'s constructor throws if it can't initialize. (At least in the case of libc++, that means on a failure to open /dev/random.)